### PR TITLE
expanding equations when passed to `@reaction_network`

### PR DESCRIPTION
**Pull request by vyudu**
Closes [#1022. ](https://github.com/SciML/Catalyst.jl/issues/1022)

### CODE REVIEW

**Comment by isaacsas** [comment]
Can you add some tests in the `dsl_options.jl` file?

File: src/dsl.jl
Comment (isaacsas) [inline src/dsl.jl:380]:
```suggestion
        push!(rxexprs.args, equation)
```

**Comment by TorkelE** [comment]
It seems like this extracts parameters from DSL expressions (checking on my phone, so have not gone through properly)? Previously we decided quite strongly against doing any such extractions, and generally keep it to a minimum with the exception of fir reactions (we could change this of course, but it would be a rather dramatic shift)

**Comment by vyudu** [comment]
Ah, I can revert that. I wanted to make 
```julia
rn = @reaction_network
    @equations D(A) ~ hill(A, v, K, n)
end
```
work without throwing an error that v is undefined, but I see that it could be complicated to allow this inference generally. What was the main reason for not letting this, out of curiosity?

**Comment by TorkelE** [comment]
The reason has to do that it can be ambiguous what is a species, a variable, and a parameter, from just looking at the expression. And if there is some misunderstanding, there could be nasty errors.

Given that equations in ReactionSystems is relatively niche, we figured it would be better to have the user manually declare in these situations.

File: test/dsl/dsl_options.jl
Comment (isaacsas) [inline test/dsl/dsl_options.jl:205]: Can we also add similar tests using a registered function for `f(A,t)`? (Just to make sure that case works too.)

Comment (isaacsas) [reply test/dsl/dsl_options.jl:205]: Ahh right, forgot about that! But maybe make one you explicitly create in case we change that in the future? (I've been debating whether we should stop registering those functions and just let MTK trace -- registration is currently breaking unit checks in them.)

Comment (vyudu) [reply test/dsl/dsl_options.jl]: As in explicitly define the `hill` function inside the test block?

Comment (isaacsas) [reply test/dsl/dsl_options.jl]: Sure, but don't call it `hill`.

**Comment by vyudu** [comment]
Hmm, escaping both sides of the equation seems to let you write more complicated LHS differential expressions a la #1044. This should be fine right? If we want users to be able to define custom `D`'s they should be resolved in their call scope anyway right?

File: test/dsl/dsl_options.jl
Comment (vyudu) [inline test/dsl/dsl_options.jl:190]: Is this what you had in mind for the registered/unregistered function test @isaacsas?

File: src/dsl.jl
Comment (TorkelE) [inline src/dsl.jl:693]: seems some stuff here got removed

**Comment by TorkelE** [comment]
This does a lot of stuff with how equations in the DSL are handled which I don't think we should touch here (e.g. permits f(d(X)) LHS). We did have some reasons to be restrictive with how we infer variables, but I really don't mind expanding this either. However, I think we should let this just handle the function errors, and then we can discuss separately which additional cases we wish to support and not, and handle those in a separate PR.

File: test/dsl/dsl_options.jl
Comment (isaacsas) [reply test/dsl/dsl_options.jl:191]: This is what I had in mind for the two ways of specifying external functions, yes. 

I'd suggest making two versions of the system with the same name (one constructed via the symbolic interface), then you can test if they are equal. That would establish that the DSL interface is generating the same system as the symbolic interface.

File: src/dsl.jl
Comment (isaacsas) [inline src/dsl.jl:729]: @vyudu are these changes needed for the function bug fix? If not, can we move them to a followup PR to expand the supported LHS's in ODEs? Let's keep this focused on just fixing the bug with the current setup.

File: src/dsl.jl
Comment (isaacsas) [reply src/dsl.jl:693]: It was refactored to a function that looks like it is trying to handle more general LHS. All of that makes sense to do as we should try to support any valid ODE that MTK can support, but that should really be a in a followup PR to generalize what we support (keeping this focused on just fixing the function bug).

**Review by isaacsas — APPROVED** [review]
This looks good to me assuming docs build ok. @TorkelE can you confirm it is ok with you before we merge?

**Comment by TorkelE** [comment]
I will have a look later today, there is a thing I want to check on my computer.

File: src/dsl.jl
Comment (TorkelE) [inline src/dsl.jl:931]: I'd probably remove the commented-out lines of code, and remove the superfluous empty lines in the function to compress it a bit. But a matter of test, up to you.

**Comment by TorkelE** [comment]
If you ctrl + f search for "### " in the dsl_options.jl test file you will find a section with tests for adding coupled equations to the DSL, I'd move the tests there. Otherwise looks good.
(we have been meaning to have a short dev doc page somewhere with stuff like this, without it is is quite hard to know about stuff likes that!)

**Pull request merged**
